// @quarryprotocol/quarry-sdk/dist/cjs/wrappers/mine/payroll.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Payroll = exports.BASE_TEN = exports.ZERO = void 0;
const tslib_1 = require("tslib");
const token_utils_1 = require("@saberhq/token-utils");
const bn_js_1 = (0, tslib_1.__importDefault)(require("bn.js"));
exports.ZERO = new bn_js_1.default(0);
exports.BASE_TEN = new bn_js_1.default(10);
class Payroll {
    constructor(famineTs, lastCheckpointTs, annualRewardsRate, rewardsPerTokenStored, totalTokensDeposited) {
        this.famineTs = famineTs;
        this.lastCheckpointTs = lastCheckpointTs;
        this.annualRewardsRate = annualRewardsRate;
        this.rewardsPerTokenStored = rewardsPerTokenStored;
        this.totalTokensDeposited = totalTokensDeposited;
    }
    /**
     * Calculates the amount of tokens that this user can receive.
     * @param current_ts
     * @returns
     */
    calculateRewardPerToken(current_ts) {
        if (this.totalTokensDeposited.isZero()) {
            return this.rewardsPerTokenStored;
        }
        const lastTimeRewardsApplicable = bn_js_1.default.min(current_ts, this.famineTs);
        const timeWorked = bn_js_1.default.max(exports.ZERO, lastTimeRewardsApplicable.sub(this.lastCheckpointTs));
        const reward = timeWorked
            .mul(new bn_js_1.default(token_utils_1.MAX_U64.toString()))
            .mul(this.annualRewardsRate)
            .div(new bn_js_1.default(365 * 86400))
            .div(this.totalTokensDeposited);
        return this.rewardsPerTokenStored.add(reward);
    }
    /**
     * Calculates the amount of tokens that this user can claim.
     * @param currentTs
     * @param tokensDeposited
     * @param rewardsPerTokenPaid
     * @param rewardsEarned
     * @returns
     */
    calculateRewardsEarned(currentTs, tokensDeposited, rewardsPerTokenPaid, rewardsEarned) {
        const netNewRewards = this.calculateRewardPerToken(currentTs).sub(rewardsPerTokenPaid);
        const earnedRewards = tokensDeposited
            .mul(netNewRewards)
            .div(new bn_js_1.default(token_utils_1.MAX_U64.toString()));
        return earnedRewards.add(rewardsEarned);
    }
}
exports.Payroll = Payroll;
